system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
   // This would be a change from the current FS.
-  if (fd == 1 || fd == 2) {
-    __wasi_size_t num = 0;
-    for (size_t i = 0; i < iovs_len; i++) {
-      const uint8_t* buf = iovs[i].buf;
-      __wasi_size_t len = iovs[i].buf_len;
-      for (__wasi_size_t j = 0; j < len; j++) {
-        uint8_t current = buf[j];
-        if (current == 0 || current == 10) {
-          fd_write_stdstream_buffer.push_back('\0'); // for null-terminated C strings
-          if (fd == 1) {
-            emscripten_console_log(&fd_write_stdstream_buffer[0]);
-          } else if (fd == 2) {
-            emscripten_console_error(&fd_write_stdstream_buffer[0]);
-          }
-          fd_write_stdstream_buffer.clear();
-        } else {
-          fd_write_stdstream_buffer.push_back(current);
-        }
+  __wasi_errno_t write(
+    const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) override {
+    return write_buffer(iovs, iovs_len, nwritten, &emscripten_console_error);
+  }
+};
+
+class OpenFile {
+  std::shared_ptr<File> file_node;
+  uint32_t _offset;
+  std::mutex openfile_mutex;
+
+public:
+  OpenFile(uint32_t offset, std::shared_ptr<File> _file_node) {
+    this->_offset = offset;
+    this->file_node = _file_node;
+  }
+
+  std::shared_ptr<File> get_file_node() { return file_node; }
+};
+
+class FileTable {
+  std::vector<std::shared_ptr<OpenFile>> handles;
+  std::mutex ft_mutex;
+
+public:
+  std::vector<std::shared_ptr<OpenFile>> get_file_handles() { return handles; }
+  std::shared_ptr<OpenFile> get_open_file(__wasi_fd_t fd) {
+    // Check if the file descriptor is invalid.
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd >= handles.size() || fd < 0)
+      return nullptr;
+
+    return handles[fd];
+  }
+
+  __wasi_fd_t add_open_file(std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    for (__wasi_fd_t i = 0; i < handles.size(); i++) {
+      if (!handles[i]) {
+        // Free open file entry.
+        handles[i] = ptr;
+        return i;
       }
-      num += len;
     }
-    *nwritten = num;
+
+    // Could not find an empty open file table entry.
+    handles.push_back(ptr);
+
+    return handles.size() - 1;
   }
-  return 0;
+
+  int remove_open_file(__wasi_fd_t fd) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    // Check if the file descriptor is invalid.
+    if (fd >= handles.size() || fd < 0)
+      return -1;
+
+    handles[fd] = nullptr;
+    return 0;
+  }
+
+  int set_open_file(__wasi_fd_t fd, std::shared_ptr<OpenFile> ptr) {
+    const std::lock_guard<std::mutex> lock(ft_mutex);
+    if (fd < 0)
+      return -1;
+
+    if (fd >= handles.size()) {
+      handles.resize(fd);
+      handles.insert(handles.begin() + fd, ptr);
+    } else {
+      handles[fd] = ptr;
+    }
+
+    return 0;
+  }
+};
+
+// Global filetable singleton
+static FileTable* ft = nullptr;
+
+void populate_file_table() {
```

Comment:
Instead of having to remember to check `ft` and call this function everywhere if it is null, how about we make a function `get_file_table` such that the only way to get a file table at all is to call that function? Then the check can be done inside that function and it will be impossible to forget to do the check or populate the table.

There are a few ways to do this, but one way would be to populate the table in the `FileTable` constructor, make the constructor private, and have a static method `FileTable::get()` that returns a reference to the file table, constructing it if necessary. You can also use a `static FileTable` local variable to lazily initialize the table without having to use `new`.